examples/timeseries/timeseries_traffic_forecasting.py

```diff
@@ -0,0 +1,655 @@
+"""
+Title: Traffic forecasting using graph neural networks and LSTM
+Author: [Arash Khodadadi](https://www.linkedin.com/in/arash-khodadadi-08a02490/)
+Date created: 2021/12/28
+Last modified: 2021/12/28
+Description: This notebook demonstrates how to do timeseries forecasting over graphs.
+"""
+"""
+## Introduction
+"""
+
+"""
+This example shows how to forecast traffic condition using graph neural networks and LSTM.
+Specifically, we are interested in predicting the future values of the traffic speed given
+a history of the traffic speed for a collection of road segments. One popular method to
+solve this problem is to consider each road segment's traffic speed as a separate
+timeseries
+and predict the future values of each timeseries using the past values of the same
+timeseries.
+This method, however, ignores the dependency of the traffic speed of one road segment on
+the neighboring segments. To be able to take into account the complex interactions between
+the traffic speed on a collection of neighboring roads, we can define the traffic network
+as a graph and consider the traffic speed as a signal on this graph. In this example,
+we implement a neural network architecture which can process timeseries data over a graph.
+We first show how to process the data and create a
+[tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) for
+forecasting over graphs. Then, we implement a model which uses graph convolution and
+LSTM layers to perform forecasting over a graph.
+
+The data processing and the model architecture are inspired by this paper:
+
+Yu, Bing, Haoteng Yin, and Zhanxing Zhu. "Spatio-temporal graph convolutional networks:
+a deep learning framework for traffic forecasting." Proceedings of the 27th International
+Joint
+Conference on Artificial Intelligence. 2018.
+([github](https://github.com/VeritasYin/STGCN_IJCAI-18))
+"""
+
+"""
+## Setup
+"""
+
+import pandas as pd
+import numpy as np
+import os
+from typing import Tuple, Optional
+import matplotlib.pyplot as plt
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers, models, optimizers, losses, callbacks, initializers
```

Comment:
One import per line. Only import modules that you use multiple times; typically only `layers` needs to be imported. Here losses, optimizers, initializers are only used once each so you can do e.g. `keras.optimizers.Adam`, etc.